### PR TITLE
Common code for assignments and default fixed moved to compiler

### PR DESCRIPTION
**Pull request by leolara**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 /target
 /Cargo.lock
+.vscode
```
Comment (qwang98) [inline .gitignore:3]: Space at the end :)

File: examples/fibonacci.rs [hunk 2]
```diff
@@ -109,6 +108,8 @@ fn fibo_circuit<F: Field + From<u64> + Hash>() -> (Circuit<F>, Option<TraceGener
             })
         });
 
+        ctx.pragma_num_steps(11);
+
```
Comment (qwang98) [inline examples/fibonacci.rs:112]: Probably should have mentioned in the last PR, but looks like `num_steps` in AST is used to calculate `num_rows` in CompilationUnit, which is used to calculate the number of assignment rows. This is different from `first_step` and `last_step`, which create a fixed column eventually.

My question is: How does `num_steps` enforce the number of steps if not through a fixed column? Can the user still pass in a witness with different number of steps while still passing the circuit constraints?

Comment (leolara) [reply examples/fibonacci.rs:112]: https://github.com/privacy-scaling-explorations/chiquito/pull/58/files#diff-19a5e6a0406f9184fd74dba07991bee321e608450fff0e30117f02c49c97ea8aR585-R586

File: .gitignore (hunk 1 above)
Comment (leolara) [reply .gitignore]: Done

File: src/wit_gen.rs [hunk 3]
```diff
@@ -100,10 +100,10 @@ pub struct FixedGenContext<F> {
 }
 
 impl<F: Field + Hash> FixedGenContext<F> {
-    pub fn new(num_steps: usize) -> Self {
+    pub fn new(num_rows: usize) -> Self {
         Self {
             assigments: Default::default(),
-            num_steps,
+            num_steps: num_rows,
```
Comment (qwang98) [inline src/wit_gen.rs:106]: Should we just call this field `num_rows` instead of `num_steps`?

File: src/compiler.rs [hunk 4]
```diff
@@ -107,11 +110,54 @@ impl<F> CompilationUnit<F> {
...
+            fixed_signals: ast.fixed_signals.clone(),
+            ..Default::default()
+        }
+    }
+}
+
+pub struct CompilerConfig<CM: CellManager, SSB: StepSelectorBuilder> {
+    cell_manager: CM,
+    step_selector_builder: SSB,
+}
```
Comment (qwang98) [inline src/compiler.rs:136]: What's the use of `CompilerConfig` as it's literally the same as `Compiler`. I only see it used once in super circuit.

File: src/wit_gen.rs (hunk 3 above)
Comment (leolara) [reply src/wit_gen.rs]: Done

File: src/compiler.rs [hunk 5]
```diff
@@ -240,43 +230,32 @@ impl<CM: CellManager, SSB: StepSelectorBuilder> Compiler<CM, SSB> {
...
+                AssigmentGenerator::new(
+                    unit.columns,
+                    unit.placement,
+                    unit.selector,
+                    TraceGenerator::new(Rc::clone(v)),
+                    unit.num_rows,
+                    uuid,
+                )
+            }),
         )
```
[the remark this answers is not in the document]
Comment (leolara) [reply src/compiler.rs:262]: step_types will never be needed, once we are in the paradigm of (ir::Column,rot) and not in the one of Queriables, all that staff is not needed, as it is needed to compile from Queriables to (ir::Column,rot)

**Comment by qwang98** [comment]
Very excited about the super circuits! Just to check my understanding, an example will be:

super_circuit: SuperCircuit = super_circuit("super circuit", |ctx: SuperCircuitContext| {
  ctx.add(compiler_config, ast_circuit_1);
  ctx.add(compiler_config, ast_circuit_2);
  ctx.mapping(|ctx: MappingContext, (u64, F, bool)| {
    ctx.map(assignment_generator_1, (u64, F)); // assignment generators are returned from compiler.compile
    ctx.map(assignment_generator_2, (F, bool));
  }
}

mapping_generator = MappingGenerator { super_circuit.mapping }
super_assignments = mapping_generator.generate((u64, F, bool));

Then, in halo2 boilerplate, call `chiquito2Halo2` on each of the ir::Circuit objects in the super circuit.

After that, in the `synthesize` function, loop through each of the `ChiquitoHalo2` objects, and call halo2::synthesize on the corresponding assignment object from super_assignments with the same uuid.

My question is: Will super circuit implement the Circuit trait from Halo2 and its `without_witness`, `configure`, `synthesize` functions?

File: src/compiler.rs (hunk 4 above)
Comment (leolara) [reply src/compiler.rs:139]: It is used here: https://github.com/privacy-scaling-explorations/chiquito/pull/58/files#diff-477cf4d3133e8c73e0a761339f1d361e646943fe2f8f5c332ccdf96df759ed66R31

I think in the end the compiler will not need to contain this fields, they will be removed.

**Comment by leolara** [comment]
> Very excited about the super circuits! Just to check my understanding, an example will be:
> 
> ```
> super_circuit: SuperCircuit = super_circuit("super circuit", |ctx: SuperCircuitContext| {
>   ctx.add(compiler_config, ast_circuit_1);
>   ctx.add(compiler_config, ast_circuit_2);
>   ctx.mapping(|ctx: MappingContext, (u64, F, bool)| {
>     ctx.map(assignment_generator_1, (u64, F)); // assignment generators are returned from compiler.compile
>     ctx.map(assignment_generator_2, (F, bool));
>   }
> }
> 
> mapping_generator = MappingGenerator { super_circuit.mapping }
> super_assignments = mapping_generator.generate((u64, F, bool));
> ```
> 
> Then, in halo2 boilerplate, call `chiquito2Halo2` on each of the ir::Circuit objects in the super circuit.
> 
> After that, in the `synthesize` function, loop through each of the `ChiquitoHalo2` objects, and call halo2::synthesize on the corresponding assignment object from super_assignments with the same uuid.
> 
> My question is: Will super circuit implement the Circuit trait from Halo2 and its `without_witness`, `configure`, `synthesize` functions?

So, in the case of halo2, for each sub_circuit there will be a different `region`.

We will need to add stuff to the halo2 backend, and some boilerplate from the examples will be moved there.

Also, we will add the possibility of adding raw halo2 sub circuits to the super circuit, but perhaps we can leave that for the future when someone needs it.

**Comment by leolara** [comment]
@qwang98 please, go again on it

File: examples/fibonacci.rs (hunk 2 above)
Comment (qwang98) [reply examples/fibonacci.rs:112]: Then I'd just comment that we are only allowing circuits with the same height for all steps, which I think we discussed last time and are fine with it for now.

File: src/compiler.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (qwang98) [reply src/compiler.rs:262]: Do you mean `place_queriable` which converts `Queriable` to `PolyExpr::Query(column, rot, annotation)`?

File: src/compiler.rs (hunk 4 above)
Comment (qwang98) [reply src/compiler.rs:139]: Kk. I saw it's used in the link. Just saying that it's the same struct as Compiler basically. Good with eventually removing them! :)

**Comment by qwang98** [comment]
> > Very excited about the super circuits! Just to check my understanding, an example will be:
> > ```
> > super_circuit: SuperCircuit = super_circuit("super circuit", |ctx: SuperCircuitContext| {
> >   ctx.add(compiler_config, ast_circuit_1);
> >   ctx.add(compiler_config, ast_circuit_2);
> >   ctx.mapping(|ctx: MappingContext, (u64, F, bool)| {
> >     ctx.map(assignment_generator_1, (u64, F)); // assignment generators are returned from compiler.compile
> >     ctx.map(assignment_generator_2, (F, bool));
> >   }
> > }
> > 
> > mapping_generator = MappingGenerator { super_circuit.mapping }
> > super_assignments = mapping_generator.generate((u64, F, bool));
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Then, in halo2 boilerplate, call `chiquito2Halo2` on each of the ir::Circuit objects in the super circuit.
> > After that, in the `synthesize` function, loop through each of the `ChiquitoHalo2` objects, and call halo2::synthesize on the corresponding assignment object from super_assignments with the same uuid.
> > My question is: Will super circuit implement the Circuit trait from Halo2 and its `without_witness`, `configure`, `synthesize` functions?
> 
> So, in the case of halo2, for each sub_circuit there will be a different `region`.
> 
> We will need to add stuff to the halo2 backend, and some boilerplate from the examples will be moved there.
> 
> Also, we will add the possibility of adding raw halo2 sub circuits to the super circuit, but perhaps we can leave that for the future when someone needs it.

Got it! Might do it as a future PR then. :)

**Comment by qwang98** [comment]
I'd just comment on the fact that fixed columns created through `fixed_gen` have the same length as the number of steps rather than the number of rows (step * step height). Just to make sure this is intended.

**Comment by leolara** [comment]
> I'd just comment on the fact that fixed columns created through `fixed_gen` have the same length as the number of steps rather than the number of rows (step * step height), which we used previously. Just to make sure this is intended. Otherwise, I'm good with everything and feel free to merge.

Yes, so I changed it to num_steps

**Review by qwang98 — APPROVED** [review]
All good!

**Pull request merged**